kvs: Support namespace prefix in key passed in by user

chu11 opened the pull request:
This is the first half of the big patches to support namespace prefixes in issue #1341.

The first patches in this series are to support parsing of namespace prefixes.  Then the support is added into the internal lookup and kvstxn APIs.  Then finally some unit tests are added.

Note that I know I did not add documentation for namespace prefixes.  I will do that later on as #1341 is completed.

chu11 commented:
Hit a

```
PASS: t0015-cron.t 21 - module load with sync
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
```

restarting that build.

chu11 commented:
re-pushed with changes discussed in #1341.  Namely, eliminate the support of "chaining" of namespaces.  By doing so, a lot of code is simplified.

Since I had to refactor a nice chunk of code in this area, I went ahead and began support of #1421 as well.  So now the namespace format of ```ns:X/``` is used.

garlick commented:
Would the strdups of chunks of key be avoidable if the fully qualified key was passed around with a "namespace length" value (0 for no namespace, number of characters in the namespace prefix o/w?  Then strncmp could be used to compare namespaces, and addition could be used to get the key suffix?

chu11 commented:
I see what you're thinking.  Some strdups could be avoided, but at several points, I need the namespace in its own string buffer, b/c it gets passed around (i.e. lookup namespace, pass namespace in rpc, etc.).  So if I didn't strdup, I'd be copying into a buffer anyways.

Perhaps minimizing the strdups is something that could be done as an optimization later on, through a alternate ```kvs_namespce_prefix_pointers()``` or something function.

On this note, I noticed one place where ```kvs_namespace_prefix()``` doesn't need to parse the key_suffix.  I went ahead and adjusted this, and squashed it since it was so tiny.

chu11 commented:
Just remembered, in ```json_pack()```, ```s#``` or ```s%``` can be used to specify a string with a length.  So that's perhaps one way to remove a strdup.

garlick commented:
That was just a suggestion - if you're happy with this as is, I'm happy.  LMK and I'll push the button.

chu11 commented:
I'm good with it.
